### ISSUE TITLE
Local indexes could allow range queries on indexing elements

### ISSUE THREAD

**Issue by psarna** [scylladb/scylla#5547]
Since the underneath structure of a local index key is
`((base_partition_key), indexed_column, base_clustering_key)`

, we could allow querying ranges from the indexed column, which could make local indexes substantially more powerful, e.g.:
```cql
CREATE TABLE t (id int PRIMARY KEY, c int, mtime int);
CREATE INDEX ON ((id), mtime); -- a local index
SELECT * FROM t WHERE id = 7 and mtime > 30;
```

Global indexing is not eligible for this, because it places the indexed column as a partition key. Local indexes could benefit from this, probably with rather cheap development cost - the infrastructure is mostly in place already.

Inspired by https://github.com/scylladb/scylla/issues/5545#issuecomment-570185690

/cc @nyh @eliransin

**Comment by slivne** [comment]
I am not sure we should not force ALLOW FILTERING in this case - SI (and LSI) there is a lookup based on the base clustering key - and that will mean we will lookup multiple rows on the same partition. 

On large partitions this double lookup will be bad - so the user should provide input that he is aware of that by adding the ALLOW FILTERING.

**Comment by psarna** [comment]
Ok, I'm absolutely fine with `ALLOW FILTERING` staying, let's just treat this issue as a possible optimization in certain cases of indexing.

**Comment by RoDmitry** [comment]
I am concerned, that `... AND v > 0 ALLOW FILTERING` is not using an index, but the index is already sorted. Why not use it? Currently it’s a full table scan of all the rows for a selected part of a PRIMARY KEY [more at forum scylladb](https://forum.scylladb.com/t/local-secondary-index-filtering/1050/3).

**Comment by RoDmitry** [comment]
```
CREATE TABLE ks.t (pk int, ck int, v1 int, v2 int, PRIMARY KEY (pk, ck)); 
CREATE INDEX ON ks.t ((pk), v1);
```
We have an index table `ks.t_v_idx_index` which gives me what I need, and I have to receive needed ids, and request the main table `ks.t`. But Scylla could do it efficiently without me sending it back (to the same node).
Maybe add extra flag to `SELECT ... USING INDEX(v)`, so to be sure that it uses index I want.

**Comment by RoDmitry** [comment]
I have noticed that `IN()` does not work on a secondary index also:
`SELECT * FROM t WHERE pk = 1 AND v1 IN(0,1);`

Which breaks my idea of sending two requests:
> We have an index table (MATERIALIZED VIEW) `ks.t_v1_idx_index` which gives me what I need, and I have to receive needed ids, and request the main table `ks.t`. But Scylla could do it efficiently without me sending it back (to the same node).

**Comment by nyh** [comment]
>  Maybe add extra flag to `SELECT ... USING INDEX(v1)`, so to be sure that it uses the index I want.

I agree that this might be a good idea. Related issues are https://github.com/scylladb/scylladb/issues/7969 (asks for "USING NO INDEX") and https://github.com/scylladb/scylladb/issues/7969.

In your case, implementing "v > 0" using the local index might be a really good thing - or a really bad thing - depending on the data:

1. If each partition has a million rows, and only **5** of them have v>0, then using the index to find the 5 will be much more efficient than scanning the million.
2. But if each partition has a million rows and **half** of them have v>0, then using the index to find them will be much, much slower than scanning the million (you'll be surprised how inefficient index scanning can be - it needs to scan the view and then read base-table partitions one by one, which is very slow).

In an ideal world, Scylla would have a clever "query planner" which, based on statistics about the data and/or how previous queries ran, can decide for each request whether to use the index or not. But since we don't have this, I agree that a way to let the user pick explicitly whether to use an index (or which index to use, if there are several options), is better than nothing.

**Comment by RoDmitry** [comment]
Agree, but you gave 2 links to the same issue :)

**Comment by fee-mendes** [comment]
> In your case, implementing "v > 0" using the local index might be a really good thing - or a really bad thing - depending on the data

Good, that's exactly what I thought after thoroughly thinking it and pointed out at [the forum](https://forum.scylladb.com/t/local-secondary-index-filtering/1050/7?u=felipemendes)

Also note https://github.com/scylladb/scylladb/pull/8667#discussion_r634180716 . I lean towards what Sarna suggested back then, given that `ORDER BY` on secondary indexes is not yet supported. Although one would most of the times expect (we hope) that a range query involving a LSI behaves in the same way as DynamoDB LSI (where the result set is going to be fetched in the indexed column clustering order), some may want the opposite: In reverse order! 

We also need to understand how to present that data? Eg: In a table `(pk, ck)` with an index on `v`, querying the `v` index as `SELECT * FROM x WHERE pk = ? AND v > ? AND v < ?` should we return the base table columns first with `ck` out of order? Or should we return the `v` column sorted first? Or should we just follow the base table sorting? What if I one cares the most on the `ck` order (ie: Expects a timestamp sorted which "simply" matches the index restrictions)?

**Comment by RoDmitry** [comment]
If I understood correctly, you mean that you select from the index table (with ordered by `v`), and then query the main table with these ids, and place it in an array (internally), which means it will be sorted by `v`. If so, then `USING INDEX(v)` would mean that it's supposed to be ordered by `v`, and only providing `ORDER BY ck` must reorder the result (maybe then instead of an `array` use `map` (internally) with key as `(pk, ck)`.

**Comment by nyh** [comment]
> We also need to understand how to present that data? Eg: In a table `(pk, ck)` with an index on `v`, querying the `v` index as `SELECT * FROM x WHERE pk = ? AND v > ? AND v < ?` should we return the base table columns first with `ck` out of order? Or should we return the `v` column sorted first? Or should we just follow the base table sorting? What if I one cares the most on the `ck` order (ie: Expects a timestamp sorted which "simply" matches the index restrictions)?

This is a very good point that I haven't considered, and I think basically shoots down this entire idea - perhaps to the point the issue should be closed as can't-be-done. Unlike a materialized view where the user knows the data is reordered, when you do a select on the base table, you expect the result to be ordered according to the original base table's clustering key - NOT the indexed key (which you don't even know if will be used or not when you're making a query on the base table). So you can't really scan the index, you'll get results in the wrong order, and if there are too many of them, you'll be out of luck.

One thing we *could* perhaps support is a query like this with **paging disabled** - Scylla could then retrieve all the results from the index (if there aren't many), re-sort them, and then return this. We actually do this already for a combination of ORDER BY and IN - and we even have bugs in that (https://github.com/scylladb/scylladb/issues/15099) :-(

**Comment by RoDmitry** [comment]
Maybe just return it ordered according to the indexed key, and specify it in the docs for `USING INDEX(v)`? Because it would be still better than me sending two read queries one after another.

**Comment by RoDmitry** [comment]
This #13533 might be related

**Comment by p-flock** [comment]
> > We also need to understand how to present that data? Eg: In a table `(pk, ck)` with an index on `v`, querying the `v` index as `SELECT * FROM x WHERE pk = ? AND v > ? AND v < ?` should we return the base table columns first with `ck` out of order? Or should we return the `v` column sorted first? Or should we just follow the base table sorting? What if I one cares the most on the `ck` order (ie: Expects a timestamp sorted which "simply" matches the index restrictions)?
> 
> This is a very good point that I haven't considered, and I think basically shoots down this entire idea - perhaps to the point the issue should be closed as can't-be-done. Unlike a materialized view where the user knows the data is reordered, when you do a select on the base table, you expect the result to be ordered according to the original base table's clustering key - NOT the indexed key (which you don't even know if will be used or not when you're making a query on the base table). So you can't really scan the index, you'll get results in the wrong order, and if there are too many of them, you'll be out of luck.
> 
> One thing we _could_ perhaps support is a query like this with **paging disabled** - Scylla could then retrieve all the results from the index (if there aren't many), re-sort them, and then return this. We actually do this already for a combination of ORDER BY and IN - and we even have bugs in that (#15099) :-(

I have come across a use case in my application / Scylla cluster that could use support of range queries on LSIs - I wouldn't personally view it as a dealbreaker if a statement using an index and a range query, e.g. `SELECT * FROM x WHERE pk = ? AND v > ? USING INDEX _` returned results out of order with respect to the base table, or even the indexed column for that matter. If the behavior is well documented (even if that means stating that there is no consistent ordering guarantee for range queries over LSIs), then users can know that if they want ordering on these results they have to do it client-side.

Seems like there can be a balance in making LSIs more powerful while acknowledging the architecture of the database prevents ideal results. Whether this feature could cause dangerous queries, e.g. an index scan being vastly less efficient than the base table scan, seems more worrying as generally Scylla holds the strong opinion that unsafe or very inefficient queries are simply not accepted besides when using `ALLOW FILTERING` - so might be a good place to require a flag, whether it be allow filtering or `USING INDEX` which at least signals the user is making an intentional choice about the query execution.

**Comment by ethanresnick** [comment]
I agree with @p-flock: if the query specifies `USING INDEX` or specifies an `ORDER BY` that references the indexed column(s), then I don't think it's problematic at all for the results to not be returned in the natural order of the base table.

**Comment by fee-mendes** [comment]
@tzach @nyh seems we have quite a few interesting suggestions and community feedback over allowing something like this. 

I believe it makes sense and this effectively makes us on pair with DDB LSIs. Thoughts?
